chainer/training/extensions/parameter_statistics.py

```diff
@@ -0,0 +1,176 @@
+        except ValueError:  # x.size == 0 and f in ('min, 'max')
+            stats[f] = float('NaN')
+
+    return stats
+
+
+def _percentiles(x, sigmas):
+    """Compute percentiles for the given array.
+
+    Args:
+        x (numpy.ndarray or cupy.ndarray): Target array for which percentiles
+            are computed.
+        sigmas (iterable or float): Percentile sigma values.
+
+    Returns:
+        numpy.ndarray or cupy.ndarray: List of percentiles. The list has the
+            same length as the given ``sigma``.
+    """
+    def _percentiles_cpu(_x):
+        if _x.size == 0:
+            return numpy.array((float('NaN'),) * len(sigmas))
+        return numpy.percentile(_x, sigmas).astype(_x.dtype)
+
+    if cuda.available and isinstance(x, cuda.ndarray):
+        # cuda.cupy.percentile() is not implemented
+        return cuda.to_gpu(_percentiles_cpu(cuda.to_cpu(x)))
+
+    return _percentiles_cpu(x)
+
+
+def _zeros(x):
+    """Count the number of zeros in the given array.
+
+    Args:
+        x (numpy.ndarray or cupy.ndarray): Target array for which sparsity is
+            computed.
+
+    Returns:
+        int: Number of zeros.
+    """
+    return x.size - cuda.get_array_module(x).count_nonzero(x)
+
+
+class ParameterStatistics(extension.Extension):
+    """Trainer extension to report parameter statistics.
+
+    The statistics are collected for a given :class:`~chainer.Link` or an
+    iterable of :class:`~chainer.Link`s. If a link contains child links, the
+    statistics are collected separately for each child.
+
+    Statistics that can be collected and reporter using the current scope are
+    minimum and maximum values, means, standard deviations, percentiles and
+    zero counts.
+
+    Args:
+        links (~chainer.Link or iterable of ~chainer.Link): Links containing
+            the parameters to observe. The link is expected to have a ``name``
+            attribute which is used as a part of a key in the report.
+        trigger: Trigger that decides when to aggregate the results and report
+            the values.
+        report_params (bool): If ``True``, include parameter statistics in the
+            report such as weight and bias values.
+        report_param_grads (bool): If ``True``, include parameter statistics
+            in the report such as weight and bias gradients.
+        count_zeros (bool): If ``True``, count the number of zero elements and
+            include those statistics in the report. Else, do not compute the
+            number of zero elements.
+        prefix (str): Prefix to prepend to the report keys.
+        targets (iterable of tuples): Target parameters and their attributes
+            to consider in this extension.
+    """
+    default_name = 'parameter_statistics'
+    priority = extension.PRIORITY_WRITER
+
+    def __init__(self, links, trigger=(1, 'epoch'), report_params=True,
+                 report_param_grads=True, count_zeros=False, prefix=None):
```

Comment:
Thanks, I'll fix that right away. Do you also think that we should pass percentiles and `count_zero` functions too in the same way (maybe using lambda by default)?